app/cdap/components/PipelineDetails/PipelineDetailsTopPanel/PipelineRunErrorDetails.tsx

```diff
@@ -286,65 +353,274 @@ export default function PipelineRunErrorDetails() {
+                        classes={
+                          err.stageName === highlightedStage ? HighlightedTrClassesOverride : {}
+                        }
+                      >
+                        <TableCell>{err.errorCategory}</TableCell>
+                        <TableCell classes={TdClassesOverride}>
+                          {renderWithLink(err.errorReason, err.supportedDocumentationUrl)}
+                        </TableCell>
+                        <TableCell>
+                          {renderWithLink(err.errorMessage, err.supportedDocumentationUrl)}
+                        </TableCell>
+                      </TableRow>
+                    ))}
+                  </TableBody>
+                </Table>
+                <ErrorImprovementMessage>
+                  {hasSurvey ? (
+                    <p>
+                      {/* TODO (CDAP-21109): Add i18n strings */}
+                      Help us improve the error classification. Raise improvement{' '}
+                      <a href="#">here</a>.
+                    </p>
+                  ) : (
+                    <p />
+                  )}
+                  <LogsButtonsContainer>
+                    <Button
+                      color="secondary"
+                      href={getDownloadLogsUrl(dataFetcher)}
+                      target="_blank"
+                      endIcon={<GetAppIcon />}
+                      data-testid={getDataTestid(`${TEST_PREFIX}.downloadLogsButton`)}
+                    >
+                      {T.translate(`${PREFIX}.downloadLogsButton`)}
+                    </Button>
+                    <Button
+                      variant="contained"
+                      color="secondary"
+                      onClick={viewLogs}
+                      data-testid={getDataTestid(`${TEST_PREFIX}.viewLogsButton`)}
+                    >
+                      {T.translate(`${PREFIX}.viewLogsButton`)}
+                    </Button>
+                  </LogsButtonsContainer>
+                </ErrorImprovementMessage>
+              </ErrorDetailsContainer>
+            </ClickAwayListener>
           )}
         </Provider>
       </PipelineRunErrorDetailsWrapper>
-      {logsOpened && <PipelineLogViewer toggleLogViewer={toggleLogs} withErrorBanner={true} />}
+      {logsOpened &&
+        (previewId ? (
+          <PreviewLogs
+            namespace={getCurrentNamespace()}
+            previewId={previewId}
+            onClose={closeLogs}
+          />
+        ) : (
+          <Provider store={PipelineDetailStore}>
+            <PipelineLogViewer toggleLogViewer={toggleLogs} withErrorBanner={true} />
+          </Provider>
+        ))}
     </ThemeWrapper>
   );
 }
+
+function ErrorDetailsBanner(props: IErrorDetailsBannerViewProps) {
+  return (
+    <Provider store={ErrorDetailsBannerUiStore}>
+      <ErrorDetailsBannerView {...props} />
+    </Provider>
+  );
+}
+
+export default function PipelineRunErrorDetails() {
```

Comment:
This file is getting pretty big. Please move each component to its own file.

Reply:
Done.